datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -1423,18 +1423,22 @@ impl LogicalPlan {
             let schema = Arc::clone(plan.schema());
             let name_preserver = NamePreserver::new(&plan);
             plan.map_expressions(|e| {
-                let original_name = name_preserver.save(&e);
-                let transformed_expr =
-                    e.infer_placeholder_types(&schema)?.transform_up(|e| {
+                let (e, has_placeholder) = e.infer_placeholder_types(&schema)?;
+                if !has_placeholder {
```

Comment:
I think the rationale may not be obvious to a future reader. Here is a suggestion:

```suggestion
                // Performance optimization: avoid NamePreserver  copy if no placeholders
                if !has_placeholder {
```

Reply:
Applied, thank you!